relay: Next iteration of refactoring (#2365)
getsentry/sentry-docs: src/docs/product/relay/operating-guidelines.mdx

```

See the [Configuration Options](/product/relay/options/) page for detailed descriptions of all available options.

## Performance Tuning

Relay has quite a lot of different [Configuration Options](/product/relay/options/), and
changing some of them can have more impact on Relay's behavior than others.

> Here we list a few relevant options that can help you to tune Relay to specific environment and workload:

Reviewer: ```suggestion
Relay provides a variety of [Configuration Options](/product/relay/options/). Changing some of the options has more impact on Relay's behavior than others. The following list identifies options that help you tune Relay to your organization's environment and workload:

Author: We would like to emphasize that this is just a small subset of the options, but you should start with them if you want to tweak Relay's performance and buffering behavior.

What do you think about this:
```suggestion
Relay provides a variety of [Configuration Options](/product/relay/options/). Changing some of the options has more impact on Relay's behavior than others. 
The following list identifies a few options you should check first when you want to tune Relay to your organization's environment and workload:
```

Reviewer: Yes, I think this works very well.

- `limits.max_concurrent_requests` (default: 100)

>   Number of concurrent requests your Relay instance can send to
>   the upstream (Sentry). If your event volume or connection latency to Sentry are high, it might be
>   beneficial to increase this value to gain additional throughput at the expense of
>   additional network connections.

Reviewer: ```suggestion
  Number of concurrent requests your Relay instance can send to
  the upstream (Sentry). If your event volume or connection latency to Sentry are high, you can increase this value to gain additional throughput, though that increase will be at the expense of additional network connections.
```

- `cache.event_buffer_size` (default: 1000)

  How many events Relay can buffer in its local queue before it starts rejecting new events.
  Increasing this value will also increase Relay's potential memory consumption when, for

>   example, there are network issues that prevent Relay from forwarding the received messages to Sentry.

Reviewer: ```suggestion
  example, network issues prevent Relay from forwarding the received messages to Sentry.
```

- `cache.event_expiry` (in seconds, default: 600)

>   How long Relay can keep buffered events in memory before dropping them. It is worth
>   increasing this value if you foresee a situtation when your Relay has to keep events in
>   memory for longer than the default value.

Reviewer: ```suggestion
  How long Relay can keep buffered events in memory before dropping them. You can increase this value if you anticipate when your Relay may need to keep events in memory for longer than the default value.
```

- `cache.project_expiry` (in seconds, default: 300)

>   To stay operational, Relay regularly fetches project configurations from the Sentry
>   upstream. This setting controls how often Relay does the fetch, and can be reduced to
>   speed up the configuration propagation. For instance, if you change data scrubbing options
>   in your project settings in Sentry, those will be propagated to your Relay instance faster.

Reviewer: ```suggestion
  To stay operational, Relay regularly fetches project configurations from the Sentry
  upstream. This setting controls how often Relay fetches that configuration. You can reduce the frequency to speed up the configuration propagation. For example, if you change data scrubbing options in your project settings in Sentry, those will be propagated to your Relay instance faster.
```

The last sentence is a bit confusing - do we mean that the settings will be propagated to Relay faster, or do we mean changing data scrubbing in project settings will speed up Relay?

Author: The data scrubbing configuration will be propagated to Relay faster.
So by reducing this value we _increase_ the frequency of those requests, e.g. fetching will be happening every minute, instead of every five minutes. It's basically the worst-case-scenario time that Relay has to wait to start using the configuration that was updated in Sentry.

What do you think about this:

```suggestion
To stay operational, Relay regularly fetches project configurations from the Sentry upstream. This setting controls how often Relay fetches that configuration. You can decrease this value to make the configuration propagation more frequent. For example, if you later change data scrubbing options in your project settings in Sentry, your Relay instance will become aware of those changes faster.
```

Reviewer: Yes, this is much more clear.

Author: Thanks!

- `cache.project_grace_period` (in seconds, default: 0)

  How long a project configuration can still be used after having expired.

>   Increasing this value might help when the upstream is unreachable (for example, in case of
>   network issues).

Reviewer: ```suggestion
  Increasing this value may help when the upstream is unreachable; for example, due to network issues.
```

## Request Routing

SDKs communicate with Sentry on a set of endpoints. Relay provides the same API to become a seamless drop-in replacement. This requires a set of endpoints to be accessible:

- `/api/<project-id>/envelope/`
- `/api/<project-id>/minidump/`
- `/api/<project-id>/security/`
- `/api/<project-id>/store/`
- `/api/<project-id>/unreal/`

Depending on the SDK or client, requests to these endpoints are performed with compressed content-encoding or chunked transfer-encoding. Depending on the infrastructure in front of Relay, please check the following HTTP headers are set correctly:

- `Host`: to the public host name of this Relay
- `X-Forwarded-For`: to the client IP address
- `X-Sentry-Auth`: to the value provided by the client

<Alert level="info">

Proxies often set a default maximum body size for requests. Especially native crash reports and attachments can exceed these limits. We recommend to configure a maximum client body size of _100MB_.

</Alert>

> Internally, Relay makes requests to the configured upstream to forward data and retrieve project configuration. Why highly advise not to constrain these requests. At present, Relay makes requests to the following endpoints for basic operation:

Reviewer: ```suggestion
Internally, Relay makes requests to the configured upstream to forward data and retrieve project configuration. We **highly recommend against** constraining these requests. At present, Relay makes requests to the following endpoints for basic operation:
```
